url74c52881d6_AndrewBourgeois_JGWCAConstants_tgz-pJ8-be_demmel_fun_jgwcaconstants_Tutorial4J8/src/be/demmel/fun/jgwcaconstants/bll/GWCAOperations.java

```diff
@@ -0,0 +1,2819 @@
+                } catch (InterruptedException ex) {
+                    // ignore
+                }
+                itemStillExists = this.getAgentExists((int) itemAgentIdAndDistance[0]);
+
+                long startDeadlockCheck2 = System.currentTimeMillis();
+                long currentlyElapsedTime2 = System.currentTimeMillis() - startDeadlockCheck2;
+
+                if (currentlyElapsedTime2 > 5000) {
+                    // try the next item
+                    break;
+                }
+
+            } while (itemStillExists);
+
+            itemsPicked++;
+        } while (itemsPicked != numberOfItems);
+    }
+
+    /**
+     * @param itemSlot
+     * @param bag
+     * @return the last modifier of the item and wchar_t* with customize text. Last modifier can be used to check which dye (model id 146) it is. If customize text = 0 then item is not customized.
+     * @throws IOException 
+     */
+    public int[] getItemLastModifier(int itemSlot, BagLocation bag) throws IOException {
+        GWCAPacket receivedGwcaPacket = sendAndReceivePacket(GWCAOperation.GET_ITEM_LAST_MODIFIER, itemSlot, bag != null ? bag.getValue() : ZERO);
+        return receivedGwcaPacket.getParamsAsIntArray();
+    }
+
+    /**
+     * @param itemId
+     * @return the last modifier of the item and wchar_t* with customize text. Last modifier can be used to check which dye (model id 146) it is. If customize text = 0 then item is not customized.
+     * @throws IOException 
+     */
+    public int[] getItemLastModifierById(int itemId) throws IOException {
+        GWCAPacket receivedGwcaPacket = sendAndReceivePacket(GWCAOperation.GET_ITEM_LAST_MODIFIER_BY_ID, itemId);
+        return receivedGwcaPacket.getParamsAsIntArray();
+    }
+
+    /**
+     * @param agentId
+     * @return
+     * @throws IOException 
+     */
+    public int[] getItemLastModifierByAgent(int agentId) throws IOException {
+        GWCAPacket receivedGwcaPacket = sendAndReceivePacket(GWCAOperation.GET_ITEM_LAST_MODIFIER_BY_AGENT, agentId);
+        return receivedGwcaPacket.getParamsAsIntArray();
+    }
+
+    /**
+     * @param mapId
+     * @param region
+     * @param language
+     * @throws IOException 
+     */
+    //TODO: enum for language
+    public void zoneMapEx(int mapId, Region region, int language) throws IOException {
+        this.prepareZoneMapEx(mapId);
+        this.zoneMapEx(region, language);
+    }
+
+    /**
+     * @param x
+     * @param y
+     * @param random
+     * @throws IOException 
+     */
+    public void moveToEx(float x, float y, int random) throws IOException {
+        LOG.debug("Moving to {} {} (and waiting until we reach it, until we're blocked or dead)");
+        int blocked = 0;
+
+        int status = this.getMapLoading();
+        this.moveEx(x, y, random);
+        pingSleep(0);
+
+        float distanceToTarget = 0;
+        do {
+            try {
+                Thread.sleep(150);
+            } catch (InterruptedException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)